### PR TITLE
#129521001 Feature user can join community 

### PR DESCRIPTION
**Pull request by andela-mochieng**
#### What does this PR do?

Creates a join community button that allows users to join public communities.
#### Description of Task to be completed?

Create a join community button that allows users to join public communities
#### How should this be manually tested?

Click on the join community button if it's a public community user is redirected to the community homepage.
#### Any background context you want to provide?

Enable users to join public communities
#### What are the relevant pivotal tracker stories?

#129521001 
#### Screenshots (if appropriate)

screen shot 2016-09-21 at 3 44 47 pm

Success message rendered when user requests to join a private community
screen shot 2016-09-26 at 1 02 11 pm

This is  the community homepage , the user is redirected to.
screen shot 2016-09-21 at 5 19 46 pm

### CODE REVIEW

File: codango/community/views.py [hunk 1]
```diff
@@ -1,9 +1,18 @@
...
+from django.views.generic.edit import CreateView
+from django.views.generic.list import ListView
+from resources.views import LoginRequiredMixin
+from .forms import CommunityForm
+from django.views.generic import View, TemplateView
+from django.contrib.auth.models import User
+from django.contrib import messages
+from community.models import CommunityMember, Community
+from django.http import HttpResponse, HttpResponseNotFound
+# Create your views here.
```
Comment (andela-cobiora) [inline codango/community/views.py]: is this comment still needed?

File: codango/codango/urls.py [hunk 2]
```diff
@@ -23,12 +23,14 @@
...
+
 urlpatterns = [
     url(r'^admin/', include(admin.site.urls)),
     url(r'^', include(account.urls)),
     url(r'^resource/', include(resources.urls)),
     url(r'^user/', include(userprofile.urls)),
     url(r'^comment/', include(comments.urls)),
+    url(r'^community/', include(community.urls)),
     url(r'^pair/', include(pairprogram.urls)),
     url(r'^community/', include(community.urls)),
```
Comment (andela-cobiora) [inline codango/codango/urls.py]: is there a difference between `line 33` and `line 35`?

File: codango/community/templates/community/community.html [hunk 3]
```diff
@@ -1 +1,39 @@
...
+    </div>
+</div>
+{% endblock sidebar %}
+
+{% block mobile-sidebar %}
+ <button type="button" class="visible-sm visible-xs" id="sidebar-mobile-link"> <i class="glyphicon glyphicon-chevron-right"></i></button>
+ <div class="visible-sm visible-xs" id="sidebar-mobile">
+    {% include "account/partials/community-menu.html" %}
+ </div>
+ {% endblock mobile-sidebar %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community.html]: remove space before the `{`

File: codango/community/templates/community/community.html [hunk 4]
```diff
@@ -1 +1,39 @@
...
+
+{% block content %}
+<div class="content-block">
+ {% if community_name %}
+<h1 style="text-align: center;">Welcome to {{ community_name.name }} Community!
+{% if community_name.private == False  %}
+{% if member == False %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community.html]: indent this `if` and the corresponding closing tags

File: codango/community/templates/community/community_list.html [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+{% load staticfiles %}
+{% cloudinary_includes %}
+
+{% block title %}Create A Community{% endblock title %}
+
+{% block sidebar %}
+
+<div class="col-sm-3 hidden-sm hidden-xs">
+    <div class="content-block">
+    {% include "account/partials/community-menu.html" %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community_list.html]: fix indentation

File: codango/community/templates/community/community_list.html [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    </div>
+    <div class="popular-block">
+    {% include "account/partials/popular.html" %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community_list.html]: fix indentation

File: codango/community/templates/community/community_member_list.html [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+{% block mobile-sidebar %}
+<button type="button" class="visible-sm visible-xs" id="sidebar-mobile-link"> <i class="glyphicon glyphicon-chevron-right"></i></button>
+<div class="visible-sm visible-xs" id="sidebar-mobile">
+    {% include "account/partials/community-menu.html" %}
+</div>
+{% endblock mobile-sidebar %}
+
+{% block content %}
+{% include "community/partials/community_member_list.html" %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community_member_list.html]: fix indentation

File: codango/community/tests/test_views.py [hunk 8]
```diff
@@ -1,10 +1,14 @@
 import time
-
+from django.db import IntegrityError
 from django.contrib.staticfiles.testing import StaticLiveServerTestCase
 from django.test.utils import setup_test_environment
 from selenium import webdriver
 from selenium.webdriver.common.keys import Keys
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: Are the above selenium imports being used?

File: codango/community/views.py [hunk 9]
```diff
@@ -42,3 +50,59 @@ def post(self, request, *args, **kwargs):
...
+                                         user=self.request.user, invitor=community.creator,
+                                         status='approved')
+            new_member.save()
+            messages.add_message(
+                request, messages.SUCCESS, 'Successfully joined ' + community.name + ' community!')
+            return redirect('/community/{}'.format(kwargs.get('community_id')))
+        else:
+            if community.private:
+                messages.add_message(
+                    request, messages.SUCCESS, 'Successfully sent a join community request!')
```
Comment (andela-cobiora) [inline codango/community/views.py]: On line 80 you indent your `kwargs` on the opening bracket ( i.e. user=self.request.user, etc.) but on here you are not. It'll be great to have one standard and have everyone stick to it. For now, I will say let us stick with the style used on line 80.

File: codango/community/views.py (hunk 1 above)
Comment (andela-mochieng) [inline codango/community/views.py]: removed

File: codango/community/views.py (hunk 9 above)
Comment (andela-mochieng) [inline codango/community/views.py]: fixed to follow pep8 rules

File: codango/codango/urls.py (hunk 2 above)
Comment (andela-mochieng) [inline codango/codango/urls.py]: removed duplicate

File: codango/community/templates/community/community.html (hunk 4 above)
Comment (andela-mochieng) [inline codango/community/templates/community/community.html]: correct indentation

File: codango/community/tests/test_views.py (hunk 8 above)
Comment (andela-mochieng) [inline codango/community/tests/test_views.py]: yes by hassan's tests

**Comment by andela-cobiora** [comment]
@andela-mochieng with hassan's merge we've got some merge conflicts, please resolve.

File: codango/community/templates/community/community_member_list.html [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+{% block mobile-sidebar %}
+<button type="button" class="visible-sm visible-xs" id="sidebar-mobile-link"> <i class="glyphicon glyphicon-chevron-right"></i></button>
+<div class="visible-sm visible-xs" id="sidebar-mobile">
+    {% include "account/partials/community-menu.html" %}
+</div>
+{% endblock mobile-sidebar %}
+
+{% block content %}
+    {% include "community/partials/community_member_list.html" %}
+{% endblock content %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/community_member_list.html]: Add line :)

File: codango/community/templates/community/partials/community_member_list.html [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+            <a href=""><h5 style="float: right">{{ member.status }}</h5></a>
+          </div>
+        </div>
+      </div>
+      <hr>
+      {% empty %}
+          You have no community members yet!
+      {% endfor %}
+</div>
+{% endblock %}
```
Comment (andela-cobiora) [inline codango/community/templates/community/partials/community_member_list.html]: add line :)

File: codango/community/tests/test_views.py [hunk 12]
```diff
@@ -1,68 +1,134 @@
...
+import json
+from django.contrib.auth.models import User
+from ..models import Community, CommunityMember
+from django.test.utils import setup_test_environment
 setup_test_environment()
 
 
-class TestCreateCommuntity(StaticLiveServerTestCase):
-    fixtures = ['users.json']
+# class TestCreateCommuntity(StaticLiveServerTestCase):
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: Did you forget to delete this block of commented out code?

File: codango/community/tests/test_views.py [hunk 13]
```diff
@@ -1,68 +1,134 @@
...
-        body = self.browser.find_element_by_tag_name('body')
-        self.assertNotIn('Create A New Community', body.text)
+        self.client = Client()
+        #  create and login user who creates communities
+        self.user = User.objects.create(username='New', password='community')
+        self.user.set_password('community')
+        self.user.save()
+        self.login = self.client.login(username='New', password='member')
+        # create public $ private communities
+        self.public_community = Community(
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: We have settled on using 
``` 
self.public_community = Community(name="Public Community",
                                  private=False,
                                  creator=self.user)
```

Like used in line `line 97`

File: codango/community/tests/test_views.py [hunk 14]
```diff
@@ -1,68 +1,134 @@
...
+            name="Public Community", private=False, creator=self.user)
+        self.public_community.save()
+        self.private_community = Community(
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: same as above here

File: codango/community/tests/test_views.py [hunk 15]
```diff
@@ -1,68 +1,134 @@
...
+            name="Private Community", private=True, creator=self.user)
+        self.private_community.save()
+        #  create the user to join the communities
+        self.user_to_join_community = User.objects.create(
+            username='Join', password='join')
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: Same as above here

File: codango/community/tests/test_views.py [hunk 16]
```diff
@@ -1,68 +1,134 @@
...
+        response = self.client.post(url, data)
+        self.assertEqual(response.status_code, 302)
+
+
+    def test_user_cant_join_public_community_twice(self):
+        """Test user can join a public community twice"""
+        self.assertTrue(self.login)
+        self.assertEqual(self.public_community.get_no_of_members(), 1)
+        try:
+            community = CommunityMember(community=self.public_community,
```
Comment (andela-cobiora) [inline codango/community/tests/test_views.py]: Indent is off a bit here

File: codango/community/views.py [hunk 17]
```diff
@@ -53,6 +62,51 @@ def get_context_data(self, **kwargs):
             raise Http404("Community does not exist")
         return context
 
+    def post(self, request, *args, **kwargs):
+        community = Community.objects.get(pk=kwargs.get('community_id'))
+        members = CommunityMember.objects.get(user)
+        context['community_name'] = Community.objects.get(
+            pk=kwargs.get('community_id'))
+        context['community_member'] = CommunityMember.objects.filter(
+            community=context['community_id']).all()
```
Comment (andela-cobiora) [inline codango/community/views.py]: Style consistently with convention

File: codango/community/views.py [hunk 18]
```diff
@@ -53,6 +62,51 @@ def get_context_data(self, **kwargs):
...
+    def post(self, request, *args, **kwargs):
+        community = Community.objects.get(pk=kwargs.get('community_id'))
+        community_members = CommunityMember.objects.filter(
+            community=community.id).all()
+        member = []
+        if community_members:
+            for members in community_members:
+                member.append(members.user.username)
+            if str(self.request.user) in member:
+                return redirect('/community/{}'.format(kwargs.get(
```
Comment (andela-cobiora) [inline codango/community/views.py]: Style consistently

**Pull request merged**